core/inc/gateway_ll.h

```diff
@@ -84,6 +99,14 @@ extern MODULE_HANDLE Gateway_LL_AddModule(GATEWAY_HANDLE gw, const GATEWAY_PROPE
 */
 extern void Gateway_LL_RemoveModule(GATEWAY_HANDLE gw, MODULE_HANDLE module);
 
+/** @brief  Registers a function to be called on a callback thread when #GATEWAY_EVENT happens
+*
+*   @param  gw         Pointer to a #GATEWAY_HANDLE to which register callback to
+*   @param  event_type Enum stating on which event should the callback be called
+*   @param  callback   Pointer to a function that will be called when the event happens
+*/
+extern void Gateway_AddEventCallback(GATEWAY_HANDLE gw, GATEWAY_EVENT event_type, GATEWAY_CALLBACK callback);
```

Comment:
@avranju Yes, it can fail. But since the whole event system is multi-threaded and to check for all the errors would need either another callback or error polling I've decided for the simplest way possible for now - disable all callbacks and log an error (see requirements doc), until there's a need for better error handling.